File: algorithm/python/LineUp2252/lineUp2252.py
import sys
from queue import Queue

input = sys.stdin.readline

N, M = map(int,input().split())

adj = [[] for _ in range(N+1)]
degree = [0] * (N+1)
queue = Queue()

for i in range(0,M):
    pre, post = map(int,input().split())

    adj[pre].append(post)
    degree[post] = degree[post] + 1

for i in range(1,N+1):
    if (degree[i] == 0):
        queue.put(i)

while not queue.empty():

    cur = queue.get()
    
    print(cur)
    
    for i in range(0,len(adj[cur])):
        next = adj[cur][i]
        degree[next] = degree[next] - 1

        if (degree[next] == 0):
            queue.put(next)

File: algorithm/c++/LineUp2252/lineUp2252.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;

int N,M;

vector<int> adj[32001];
int degree[32001];

queue<int> que;
queue<int> result_que;

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    
    cin >> N >> M;
    
    for (int i = 0; i < M; i++)
    {
        int pre, post;
        cin >> pre >> post;

        adj[pre].push_back(post);
        degree[post]++;
    }
    
    for (int i = 1; i <= N; i++)
    {
        if (degree[i] == 0) 
        {
            que.push(i);
            result_que.push(i);
        }
    }
    
    while(!que.empty())
    {
        int cur = que.front();
        que.pop();

        for (int i = 0; i < adj[cur].size(); i++) 
        {
            int next = adj[cur][i];
            degree[next]--;

            if (degree[next] == 0) 
            {
                que.push(next);
                result_que.push(next);
            }
        }
    }
    
    while(!result_que.empty())
    {
        int result = result_que.front();
        result_que.pop();

        cout << result << " ";
    }


    return 0;
}

